src/order/complete_well_founded.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2020 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import order.well_founded
+import order.order_iso_nat
+import data.set.finite
+
+/-!
+# Well-foundedness for complete lattices
+
+For complete lattices, there are numerous equivalent ways to express the fact that the relation `>`
+is well-founded. In this file we define two especially-useful characterisations and provide
+proofs that they are indeed equivalent to well-foundedness.
+
+## Main definitions
+ * `is_sup_closed_compact`
+ * `is_Sup_finite_compact`
+
+## Main results
+ * `is_sup_closed_compact_iff_is_Sup_finite_compact`
+ * `well_founded_iff_is_sup_closed_compact`
+
+## Tags
+
+complete lattice, well-founded, compact
+-/
+
+namespace complete_lattice
+
+variables (α : Type*) [complete_lattice α]
+
+/-- A compactness property for a complete lattice is that any `sup`-closed non-empty subset
+contains its `Sup`. -/
+def is_sup_closed_compact :=
+  ∀ (s : set α) (h : s.nonempty), (∀ a b, a ∈ s → b ∈ s → a ⊔ b ∈ s) → (Sup s) ∈ s
+
+/-- A compactness property for a complete lattice is that any subset has a finite subset with the
+same `Sup`. -/
+def is_Sup_finite_compact := ∀ (s : set α), ∃ (t ⊆ s) (h : t.finite), Sup s = Sup t
```

Comment:
Should there be another version with `t : finset α` and `finset.sup` instead of `Sup`?

Reply:
Actually, I'd probably only use the finset version.